src/PhpSpec/ServiceContainer/IndexedServiceContainer.php

```diff
@@ -0,0 +1,238 @@
+     * @param array  $tags
+     *
+     * @throws \InvalidArgumentException if service is not an object or callable
+     */
+    public function set($id, $value, $tags=[])
+    {
+        if (!is_object($value)) {
+            throw new InvalidArgumentException(sprintf(
+                'Service should be an object, but %s given.',
+                gettype($value)
+            ));
+        }
+
+        $this->addToIndex(
+            $id,
+            function() use ($value) {
+                return $value;
+            },
+            $tags
+        );
+    }
+
+    /**
+     * Sets a callable for the service creation. The same service will
+     * be returned every time
+     *
+     * @param string   $id
+     * @param callable $callable
+     * @param array    $tags
+     *
+     * @throws \InvalidArgumentException if service is not a callable
+     */
+    public function define($id, $callable, $tags = [])
+    {
+        if (!is_callable($callable)) {
+            throw new InvalidArgumentException(sprintf(
+                'Service should be callable, "%s" given.',
+                gettype($callable)
+            ));
+        }
+
+        $this->addToIndex(
+            $id,
+            function ($container) use ($callable) {
+                static $instance;
+
+                if (null === $instance) {
+                    $instance = call_user_func($callable, $container);
+                }
+
+                return $instance;
+            },
+            $tags
+        );
+    }
+
+    /**
+     * Retrieves a service from the container
+     *
+     * @param string $id
+     *
+     * @return object
+     *
+     * @throws \InvalidArgumentException if service is not defined
+     */
+    public function get($id)
+    {
+        if (!array_key_exists($id, $this->services)) {
+            throw new InvalidArgumentException(sprintf('Service "%s" is not defined.', $id));
+        }
+
+        return call_user_func($this->services[$id], $this);
+    }
+
+    /**
+     * Determines whether a service is defined
+     *
+     * @param $id
+     * @return bool
+     */
+    public function has($id)
+    {
+        return array_key_exists($id, $this->services);
+    }
+
+    /**
+     * Removes a service from the container
+     *
+     * @param string $id
+     *
+     * @throws \InvalidArgumentException if service is not defined
+     */
+    public function remove($id)
+    {
+        if (!array_key_exists($id, $this->services)) {
+            throw new InvalidArgumentException(sprintf('Service "%s" is not defined.', $id));
+        }
+
+        unset($this->services[$id]);
+    }
+
+    /**
+     * Adds a service or service definition to the index
+     *
+     * @param string $id
+     * @param mixed  $value
+     * @param array  $tags
+     */
+    private function addToIndex($id, $value, array $tags)
+    {
+        foreach ($tags as $tag) {
+            if (!isset($this->tags[$tag])) {
+                $this->tags[$tag] = [];
```

Comment:
reading triggers a warning. but ``$this->tags[$tag][] = $id;`` does not need reading the current value